python/uds.py

```diff
@@ -376,39 +378,46 @@ def send(self, dat: bytes) -> None:
     self.rx_done = False
 
     if self.debug:
-      print(f"ISO-TP: REQUEST - 0x{bytes.hex(self.tx_dat)}")
+      print(f"ISO-TP: REQUEST - {hex(self._can_client.tx_addr)} 0x{bytes.hex(self.tx_dat)}")
     self._tx_first_frame()
 
   def _tx_first_frame(self) -> None:
     if self.tx_len < self.max_len:
       # single frame (send all bytes)
       if self.debug:
-        print("ISO-TP: TX - single frame")
+        print(f"ISO-TP: TX - {hex(self._can_client.tx_addr)} single frame")
       msg = (bytes([self.tx_len]) + self.tx_dat).ljust(self.max_len, b"\x00")
       self.tx_done = True
     else:
       # first frame (send first 6 bytes)
       if self.debug:
-        print("ISO-TP: TX - first frame")
+        print(f"ISO-TP: TX - {hex(self._can_client.tx_addr)} first frame")
       msg = (struct.pack("!H", 0x1000 | self.tx_len) + self.tx_dat[:self.max_len - 2]).ljust(self.max_len - 2, b"\x00")
     self._can_client.send([msg])
 
   def recv(self) -> Optional[bytes]:
     start_time = time.time()
+    self.rx_idx_prev = 0
     try:
       while True:
         for msg in self._can_client.recv():
           self._isotp_rx_next(msg)
           if self.tx_done and self.rx_done:
             return self.rx_dat
+          if self.rx_idx > self.rx_idx_prev:
+            start_time = time.time()
```

Comment:
I was thinking we can just reset `start_time` every time a message is received (no need to add anything new tracking previous idx), is there a reason we shouldn't do that (I believe we are guaranteed `self._can_client.recv()` returning something guarantees a relevant message came in)?

Reply:
At the time I assumed `_can_client.recv()` would hand back whatever was on the wire and depend on `_isotp_rx_next()` to check for relevance and sanity. I never really checked that assumption. If you're sure we can rely on `_can_client.recv()` to only return messages of relevance to our diagnostic conversation, the `prev` stuff can be elided.

Reply:
I reviewed how it works (`CanClient` is pretty simple) and came to the above conclusion.  If I missed something let me know!

Reply:
done, tested and worked for my usecase